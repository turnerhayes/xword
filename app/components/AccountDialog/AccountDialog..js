import React      from "react";
import PropTypes  from "prop-types";
import { Link }   from "react-router-dom";
import IconButton from "@material-ui/core/IconButton";
import Icon       from "@material-ui/core/Icon";
import UserRecord from "@app/scripts/records/user";
import {
	login,
	logout
}                 from "@app/actions";
import Config     from "@app/scripts/config";
import                 "./AccountDialog.less";


function handleLogoutButtonClicked({ dispatch }) {
	dispatch(logout());
}

function handleLoginClicked({ dispatch, provider }) {
	dispatch(login({ provider }));
}

function renderNotLoggedIn(dispatch) {
	return (
		<div>
			{
				Config.auth.facebook.isEnabled && (
					<IconButton
						className="login-link"
						title="Log in with Facebook"
						aria-label="Log in with Facebook"
						onClick={() => handleLoginClicked({ dispatch, provider: "facebook" })}
					>
						<Icon
							className="icon"
						>facebook</Icon>
					</IconButton>
				)
			}
			{
				Config.auth.google.isEnabled && (
					<IconButton
						className="login-link"
						title="Log in with Google"
						aria-label="Log in with Google"
						onClick={() => handleLoginClicked({ dispatch, provider: "google" })}
					>
						<Icon
							className="icon"
						>google plus</Icon>
					</IconButton>
				)
			}
			{
				Config.auth.twitter.isEnabled && (
					<IconButton
						className="login-link"
						title="Log in with Google"
						aria-label="Log in with Google"
						onClick={() => handleLoginClicked({ dispatch, provider: "twitter" })}
					>
						<Icon
							className="icon"
						>twitter</Icon>
					</IconButton>
				)
			}
		</div>
	);
}

/**
 * Generates a React component representing the dialog shown to allow users to manage their site account
 * (log in/out, edit profile, etc.).
 *
 * @memberof client.react-components
 *
 * @param {object} props - props for the component
 *
 * @return {external:React.Component} the component to render
 */
function AccountDialog(props) {
	return (
		<div
			className="c_account-dialog"
		>
			{
				<h4
				>
					{
						props.loggedInUser && (
							<Link
								to={`/profile/${props.loggedInUser.username}`}
							>
								<Icon
									className="icon provider-icon"
								>{props.loggedInUser.provider}</Icon>
								{ props.loggedInUser.name.get("display") }
							</Link>
						)
					}
					{
						props.loggedInUser && (
							<IconButton
								aria-label="Log out"
								title="Log out"
								onClick={handleLogoutButtonClicked}
							>
								<Icon
									className="icon"
								>log out</Icon>
							</IconButton>
						)
					}
					{
						!props.loggedInUser && "Log in"
					}
				</h4>
			}
			{
				!props.loggedInUser && renderNotLoggedIn(props.dispatch)
			}
		</div>
	);
}

/**
 * Component prop types
 *
 * @prop {function} dispatch - function to dispatch actions to the Redux store
 * @prop {client.records.UserRecord} [loggedInUser] - the currently logged in user, if any
 */
AccountDialog.propTypes = {
	dispatch: PropTypes.func.isRequired,
	loggedInUser: PropTypes.instanceOf(UserRecord),
};

export default AccountDialog;
