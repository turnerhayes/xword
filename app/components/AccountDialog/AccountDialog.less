.c_account-dialog {
	.login-link {
		font-size: 2em;

		&,
		&:hover,
		&:active {
			text-decoration: none;
		}
	}

	.provider-icon {
		margin-right: 0.5em;
	}
}
