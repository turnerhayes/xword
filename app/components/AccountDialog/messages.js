import { defineMessages } from "react-intl";

export default defineMessages({
	actions: {
		logIn: {
			id: "xword.general.actions.logIn",
		},
		logOut: {
			id: "xword.general.actions.logOut",
		},
		logInWith: {
			id: "xword.components.AccountDialog.actions.logInWith",
		},
	},
});
