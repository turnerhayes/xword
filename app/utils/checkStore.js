import invariant from "invariant";

/**
 * Validate the shape of redux store
 */
export default function checkStore(store) {
	invariant(
		store,
		"(@app/utils/checkStore) injectors: Expected a store object"
	);
	
	invariant(
		typeof store.dispatch === "function",
		"(@app/utils/checkStore) injectors: Expected a dispatch function"
	);

	invariant(
		typeof store.subscribe === "function",
		"(@app/utils/checkStore) injectors: Expected a subscribe function"
	);

	invariant(
		typeof store.getState === "function",
		"(@app/utils/checkStore) injectors: Expected a getState function"
	);

	invariant(
		typeof store.replaceReducer === "function",
		"(@app/utils/checkStore) injectors: Expected a replaceReducer function"
	);

	invariant(
		typeof store.runSaga === "function",
		"(@app/utils/checkStore) injectors: Expected a runSaga function"
	);
}
