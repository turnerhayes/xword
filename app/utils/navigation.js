export const navigateToUrl = (url) => {
	document.location.assign(url);
};
