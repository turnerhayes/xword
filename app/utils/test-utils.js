import React from "react";
import PropTypes from "prop-types";
import configureStore from "redux-mock-store";
import { runSaga as realRunSaga } from "redux-saga";
import { IntlProvider } from "react-intl";
import { mount as enzymeMount, shallow as enzymeShallow } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Renderer from "react-test-renderer";

import { translationMessages } from "@app/i18n";
import createReducer from "@app/reducers";

const locale = "en";
const defaultLocale = "en";

const _mockStore = configureStore();

export const mockStore = (initialState, ...args) => {
	if (!initialState) {
		initialState = createReducer()(undefined, {});
	}

	const store = _mockStore(initialState, ...args);

	store.runSaga = () => {};

	return store;
};

const WrapperComponent = ({ children, store }) => (
	<MemoryRouter>
		<Provider store={store}>
			<IntlProvider
				locale={locale}
				defaultLocale={defaultLocale}
				messages={translationMessages[locale]}
			>
				{children}
			</IntlProvider>
		</Provider>
	</MemoryRouter>
);

WrapperComponent.propTypes = {
	children: PropTypes.any,
	store: PropTypes.object,
};

export const mount = (node, {
	store = mockStore(),
} = {}) => enzymeMount(node, {
	wrappingComponent: WrapperComponent,
	wrappingComponentProps: {
		store,
	},
});

export const shallow = (node, {
	store = mockStore(),
} = {}) => enzymeShallow(node, {
	wrappingComponent: WrapperComponent,
	wrappingComponentProps: {
		store
	},
});

export const renderComponent = (node, {
	store = mockStore(),
} = {}) => Renderer.create(
	<WrapperComponent
		store={store}
	>
		{node}
	</WrapperComponent>
);

export async function runSaga({
	state,
	getSaga,
}, ...args) {
	const dispatched = [];

	const rootSaga = await getSaga();

	const dispatchers = [];

	const reducer = createReducer();

	state = state || reducer(undefined, {});

	const getDispatch = (pushToArray = true) => {
		return (action) => {
			if (pushToArray) {
				dispatched.push(action);
			}

			state = reducer(state, action);
		};
	};

	const dispatchWithoutPushing = getDispatch(false);

	const sagaPromise = realRunSaga(
		{
			dispatch: getDispatch(),
			getState: () => state,
			subscribe: (callback) => {
				const wrappedCallback = (action, ...args) => {
					dispatchWithoutPushing(action);
					return callback(action, ...args);
				};

				dispatchers.push(wrappedCallback);

				return () => {
					const index = dispatchers.findIndex(wrappedCallback);

					if (index >= 0) {
						delete dispatchers[index];
					}
				};
			},
		},
		rootSaga,
		...args
	).done;

	return {
		dispatchers,
		dispatched,
		sagaPromise,
	};
}
