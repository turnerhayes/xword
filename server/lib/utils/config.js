"use strict";

let path   = require("path");
let fs     = require("fs");
let _      = require("lodash");
let assert = require("assert");

let staticPath = path.resolve(__dirname, "..", "..", "..", "static");
let templatesPath = path.join(staticPath, "templates");
let partialsPath = path.join(templatesPath, "partials");

let Config = {
	app: {
		environment: process.env.NODE_ENV || "development",
		secret: process.env.CONFIG_APP_SECRET,
		address: {
			host: process.env.CONFIG_APP_ADDRESS_HOST,
			port: process.env.PORT,
			isSecure: !!process.env.CONFIG_APP_ADDRESS_IS_SECURE,
			externalPort: process.env.CONFIG_APP_ADDRESS_EXTERNAL_PORT || null,
			insecurePort: process.env.CONFIG_APP_ADDRESS_INSECURE_PORT || null
		}
	},
	static: {
		host: process.env.CONFIG_STATIC_HOST || null
	},
	session: {
		key: process.env.CONFIG_SESSION_KEY,
		store: {
			url: process.env.CONFIG_SESSION_STORE_URL
		}
	},
	data: {
		store: {
			url: process.env.CONFIG_DATA_STORE_URL
		}
	},
	paths: {
		static: staticPath,
		templates: templatesPath,
		partials: partialsPath
	},
	authentication: {
		facebook: {
			profileFields: [
				"id",
				"name",
				"displayName",
				"emails"
			],
			scope: [
				"email"
			],
			callbackURL: "/auth/fb/callback"
		},
		google: {
			scope: ["https://www.googleapis.com/auth/plus.login", "email"],
			callbackURL: "/auth/google/callback"
		}
	},
	credentials: {
		facebook: {
			appSecret: process.env.CONFIG_CREDENTIALS_FACEBOOK_APP_SECRET,
			appID: process.env.CONFIG_CREDENTIALS_FACEBOOK_APP_ID
		},
		google: {
			clientSecret: process.env.CONFIG_CREDENTIALS_GOOGLE_CLIENT_SECRET,
			clientID: process.env.CONFIG_CREDENTIALS_GOOGLE_CLIENT_ID
		}
	},
	ssl: {
		key: _.isUndefined(process.env.CONFIG_SSL_KEY) ?
			(
				_.isString(process.env.CONFIG_SSL_KEY_PATH) ?
					fs.readFileSync(process.env.CONFIG_SSL_KEY_PATH).toString() :
					undefined
			) :
			process.env.CONFIG_SSL_KEY,
		cert: _.isUndefined(process.env.CONFIG_SSL_CERT) ?
			(
				_.isString(process.env.CONFIG_SSL_CERT_PATH) ?
					fs.readFileSync(process.env.CONFIG_SSL_CERT_PATH).toString() :
					undefined
			) :
			process.env.CONFIG_SSL_CERT
	}
};

const externalPort = Config.app.address.externalPort || Config.app.address.port;

Object.defineProperties(
	Config.app,
	{
		baseURL: {
			value: "http" + (Config.app.address.isSecure ? "s" : "") + "://" +
				Config.app.address.host + (
					(externalPort + "") === "80" ?
						"" :
						":" + externalPort
				)
		}
	}
);

assert(
	!Config.app.address.isSecure || (Config.ssl.key && Config.ssl.cert),
	"Both `ssl.key` and `ssl.cert` are required to run a secure server"
);

exports = module.exports = Config;
