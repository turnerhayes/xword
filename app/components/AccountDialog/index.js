import AccountDialog from "./AccountDialog";

export default AccountDialog;
