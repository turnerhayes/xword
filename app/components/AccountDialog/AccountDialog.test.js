import React from "react";
import { fromJS } from "immutable";
import Icon from "@material-ui/core/Icon";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import { mount, renderComponent } from "@app/utils/test-utils";

import AccountDialog from "./AccountDialog";

const NO_OP = () => {};

describe("AccountDialog component", () => {
	const loginLinkClass = "login-link";

	it("should show a login link for each supported provider if no user is logged in", () => {
		const providers = [ "facebook", "google", "twitter" ];

		const component = renderComponent(
			<AccountDialog
				onLogin={NO_OP}
				onLogout={NO_OP}
				enabledProviders={providers}
				classes={{}}
			/>
		);

		expect(component.toJSON()).toMatchSnapshot();
	});

	it("should not show a login link for a provider if that provider is not supported", () => {
		const component = renderComponent(
			<AccountDialog
				onLogin={NO_OP}
				onLogout={NO_OP}
				enabledProviders={[ "facebook" ]}
				classes={{}}
			/>
		);

		expect(component.toJSON()).toMatchSnapshot();
	});

	it("should call login callback with the appropriate provider when login button is clicked", () => {
		const onLogin = jest.fn();

		const selectedProvider = "facebook";

		const wrapper = mount(
			<AccountDialog
				onLogin={onLogin}
				onLogout={NO_OP}
				enabledProviders={[ selectedProvider, "twitter" ]}
				classes={{
					loginLink: loginLinkClass,
				}}
			/>
		);

		const loginButton = wrapper.find(IconButton).filterWhere(
			(el) => el.hasClass(loginLinkClass) && el.key() === selectedProvider
		);

		expect(loginButton).toExist();

		loginButton.simulate("click");

		expect(onLogin).toHaveBeenCalledWith({ provider: selectedProvider });
	});

	it("should show user display name if a user is logged in", () => {
		const displayName = "Test Tester";

		const loggedInUser = fromJS({
			name: {
				display: displayName,
			},
			username: "tester@example.com",
			provider: "facebook",
		});

		const component = renderComponent(
			<AccountDialog
				onLogin={NO_OP}
				onLogout={NO_OP}
				loggedInUser={loggedInUser}
				classes={{}}
			/>
		);

		expect(component.toJSON()).toMatchSnapshot();
	});

	it("should call logout callback when logout button clicked", () => {
		const displayName = "Test Tester";

		const loggedInUser = fromJS({
			name: {
				display: displayName,
			},
			provider: "facebook",
		});

		const onLogout = jest.fn();

		const wrapper = mount(
			<AccountDialog
				onLogin={NO_OP}
				onLogout={onLogout}
				loggedInUser={loggedInUser}
				classes={{}}
			/>
		);

		const logoutButton = wrapper.find(Button).filterWhere(
			(button) => button.findWhere(
				(el) => el.is(Icon) && el.text() === "log out"
			)
		);

		expect(logoutButton).toExist();

		logoutButton.simulate("click");

		expect(onLogout).toHaveBeenCalled();
	});
});
