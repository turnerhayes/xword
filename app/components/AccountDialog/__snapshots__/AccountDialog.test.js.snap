// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDialog component should not show a login link for a provider if that provider is not supported 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Log in
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    <div>
      <button
        aria-label="Log in with Facebook"
        className="MuiButtonBase-root MuiIconButton-root AccountDialog-loginLink-1"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Log in with Facebook"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root icon"
          >
            facebook
          </span>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`AccountDialog component should show a login link for each supported provider if no user is logged in 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Log in
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    <div>
      <button
        aria-label="Log in with Facebook"
        className="MuiButtonBase-root MuiIconButton-root AccountDialog-loginLink-1"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Log in with Facebook"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root icon"
          >
            facebook
          </span>
        </span>
      </button>
      <button
        aria-label="Log in with Google+"
        className="MuiButtonBase-root MuiIconButton-root AccountDialog-loginLink-1"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Log in with Google+"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root icon"
          >
            google plus
          </span>
        </span>
      </button>
      <button
        aria-label="Log in with Twitter"
        className="MuiButtonBase-root MuiIconButton-root AccountDialog-loginLink-1"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Log in with Twitter"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root icon"
          >
            twitter
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`AccountDialog component should show user display name if a user is logged in 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        <div>
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root icon AccountDialog-providerIcon-2"
          >
            facebook
          </span>
          <a
            href="/profile/tester@example.com"
            onClick={[Function]}
          >
            Test Tester
          </a>
        </div>
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  />
  <div
    className="MuiCardActions-root MuiCardActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root icon AccountDialog-buttonIcon-3"
        >
          log out
        </span>
        Log out
      </span>
    </button>
  </div>
</div>
`;
